core/engine/src/object/builtins/lazy_builtin.rs

```diff
@@ -0,0 +1,260 @@
+use super::JsFunction;
+use crate::{
+    gc::custom_trace,
+    object::{
+        internal_methods::{
+            non_existant_call, non_existant_construct, ordinary_define_own_property,
+            ordinary_delete, ordinary_get, ordinary_get_own_property, ordinary_get_prototype_of,
+            ordinary_has_property, ordinary_is_extensible, ordinary_own_property_keys,
+            ordinary_prevent_extensions, ordinary_set, ordinary_set_prototype_of, ordinary_try_get,
+            CallValue, InternalMethodContext, InternalObjectMethods,
+        },
+        JsPrototype,
+    },
+    property::{PropertyDescriptor, PropertyKey},
+    realm::Realm,
+    Context, JsData, JsObject, JsResult, JsValue,
+};
+use boa_gc::{Finalize, Trace};
+use std::cell::Cell;
+
+#[derive(Debug, Clone, Trace, Finalize)]
+pub(crate) enum BuiltinKind {
+    Constructor(JsFunction),
+    Ordinary,
+}
+
+/// A builtin function. Used for lazy initialization of builtins.
+
+#[derive(Debug, Clone, Finalize)]
+pub(crate) struct Builtin {
+    pub init: fn(&Realm),
+    pub is_initialized: Cell<bool>,
+    pub kind: BuiltinKind,
+}
+
+// SAFETY: Temporary, TODO move back to derived Trace when possible
+unsafe impl Trace for Builtin {
+    custom_trace!(this, mark, {
+        mark(&this.kind);
+    });
+}
+
+// Implement the trait for JsData by overriding all internal_methods by calling init before calling into the underlying internel_method
+impl JsData for Builtin {
+    fn internal_methods(&self) -> &'static InternalObjectMethods {
+        &LAZY_INTERNAL_METHODS
+    }
+}
+
+pub(crate) static LAZY_INTERNAL_METHODS: InternalObjectMethods = InternalObjectMethods {
+    __get_prototype_of__: lazy_get_prototype_of,
+    __set_prototype_of__: lazy_set_prototype_of,
+    __is_extensible__: lazy_is_extensible,
+    __prevent_extensions__: lazy_prevent_extensions,
+    __get_own_property__: lazy_get_own_property,
+    __define_own_property__: lazy_define_own_property,
+    __has_property__: lazy_has_property,
+    __try_get__: lazy_try_get,
+    __get__: lazy_get,
+    __set__: lazy_set,
+    __delete__: lazy_delete,
+    __own_property_keys__: lazy_own_property_keys,
+    __call__: lazy_call,
+    __construct__: lazy_construct,
+};
+
+pub(crate) fn lazy_get_prototype_of(
+    obj: &JsObject,
+    context: &mut Context,
+) -> JsResult<JsPrototype> {
+    let realm = context.realm();
```

Comment:
`realm` cannot be fetched here, because the initializer could be called from a different realm, which would initialize the incorrect intrinsic. We can use `Gc::new_cyclic` to store a weak reference to the `Realm` inside the builtin, then recover the strong reference inside here.